packages/hierarchies-react/src/presentation-hierarchies-react/itwinui/TreeRenderer.tsx

```diff
@@ -46,20 +45,43 @@ export function TreeRenderer({ rootNodes, expandNode, localizedStrings, selectNo
     selectionMode: selectionMode ?? "single",
   });
 
+  const flatNodes = useMemo(() => getFlatNodes(rootNodes, 0), [rootNodes]);
+
   return (
     <LocalizationContextProvider localizedStrings={localizedStrings}>
       <Tree.Root style={{ height: "100%", width: "100%" }}>
-        <TreeLevelRenderer
-          {...treeProps}
-          nodes={rootNodes}
-          expandNode={expandNode}
-          onNodeClick={onNodeClick}
-          onNodeKeyDown={onNodeKeyDown}
-          isNodeSelected={isNodeSelected}
-        />
+        {flatNodes.map((flatNode) => (
+          <TreeNodeRenderer
+            {...treeProps}
+            expandNode={expandNode}
+            onNodeClick={onNodeClick}
+            onNodeKeyDown={onNodeKeyDown}
+            node={flatNode}
+            key={flatNode.id}
+            selected={isNodeSelected?.(flatNode.id)}
+          />
+        ))}
       </Tree.Root>
     </LocalizationContextProvider>
   );
 }
 
 function noopSelectNodes() {}
+
+export type FlatPresentationTreeNode = {
+  level: number;
+  levelSize: number;
+  posInLevel: number;
+} & PresentationTreeNode;
+
+export function getFlatNodes(nodes: PresentationTreeNode[], level: number) {
```

Comment:
I think for public facing API it would be better to have something like:

```ts
/** @alpha */
export function flattenNodes(nodes: PresentationTreeNode[]) {
  return getFlatNodes(nodes, 0);
}

/** @internal */
function getFlatNodes(nodes: PresentationTreeNode[], level: number) {
}
```

Reply:
Changed.